pricegraph/src/orderbook.rs

```diff
@@ -166,37 +166,36 @@ impl Orderbook {
 
         let (base, quote) = (node_index(market.base), node_index(market.quote));
 
-        while let Err(cycle) = bellman_ford::search(&self.projection, quote) {
-            let cycle_not_in_path = match cycle.with_starting_node(quote) {
-                Err(cycle) => cycle,
-                Ok(cycle) => {
-                    let base_index_search = cycle.0.iter().position(|node| *node == base);
-                    match base_index_search {
-                        None => cycle,
-                        Some(base_index) => {
-                            let (ask, bid) = (&cycle.0[0..base_index + 1], &cycle.0[base_index..]);
-
-                            let ask = self
-                                .fill_path(ask)
-                                .expect("ask transitive path not found after being detected");
-                            let bid = self
-                                .fill_path(bid)
-                                .expect("bid transitive path not found after being detected");
-
-                            return Some(Ring { ask, bid });
-                        }
-                    }
+        while let Err(cycle) = ShortestPathGraph::new(&self.projection, quote) {
+            let paths_base_quote = match cycle.with_starting_node(quote) {
+                Err(cycle) => Err(cycle),
+                Ok(cycle) => match cycle.iter().position(|node| *node == base) {
+                    None => Err(cycle),
+                    Some(base_index) => Ok(cycle.split_at(base_index)),
```

Comment:
split_at could take a node instead of an index to avoid the extra match arm, but that's fine too.